File: java-example/src/test/java/org/onflow/examples/java/createAccount/CreateAccountExampleTest.java
package org.onflow.examples.java.createAccount;

import com.google.common.io.BaseEncoding;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.onflow.examples.java.AccessAPIConnector;
import org.onflow.flow.common.test.*;
import org.onflow.flow.sdk.FlowAccessApi;
import org.onflow.flow.sdk.FlowAddress;
import org.onflow.flow.sdk.SignatureAlgorithm;
import org.onflow.flow.sdk.crypto.Crypto;
import org.onflow.flow.sdk.crypto.KeyPair;
import org.onflow.flow.sdk.crypto.PublicKey;
import org.onflow.flow.sdk.FlowAccountKey;

import java.math.BigDecimal;

@FlowEmulatorProjectTest(flowJsonLocation = "../flow/flow.json")
public class CreateAccountExampleTest {

    private final KeyPair[] userKeyPairs = {
            Crypto.generateKeyPair(SignatureAlgorithm.ECDSA_P256),
            Crypto.generateKeyPair(SignatureAlgorithm.ECDSA_SECP256k1)
    };

    private final FlowAddress emptyAddress = new FlowAddress("");

    // user account addresses
    private final FlowAddress[] userAccountAddress = {
            emptyAddress,
            emptyAddress
    };

    @FlowServiceAccountCredentials
    private TestAccount serviceAccount;

    @FlowTestClient
    private FlowAccessApi accessAPI;

    // create an account using the service account
    private FlowAddress createUserAccount(PublicKey userPublicKey) {
        CreateAccountExample createAccountExample = new CreateAccountExample(serviceAccount.getPrivateKey(), accessAPI);
        return createAccountExample.createAccount(serviceAccount.getFlowAddress(), userPublicKey);
    }

    private String bytesToHex(byte[] data) {
        return BaseEncoding.base16().lowerCase().encode(data);
    }

    @Test
    public void canCreateAnAccount() {
        for (int index = 0; index < userAccountAddress.length; index++) {
            FlowAddress address = userAccountAddress[index];
            if (address.equals(emptyAddress)) {
                // create test accounts
                userAccountAddress[index] = createUserAccount(userKeyPairs[index].getPublic());
            }
        }

        for (int index = 0; index < userAccountAddress.length; index++) {
            FlowAddress address = userAccountAddress[index];
            Assertions.assertNotNull(address);
            // check account key is the expected one
            AccessAPIConnector accessAPIConnector = new AccessAPIConnector(serviceAccount.getPrivateKey(), accessAPI);
            FlowAccountKey newAccountKey = accessAPIConnector.getAccountKey(address, 0);
            Assertions.assertEquals(userKeyPairs[index].getPublic().getHex(), bytesToHex(newAccountKey.getPublicKey().getBytes()));
        }
    }

    @Test
    public void canGetAccountBalance() {
        AccessAPIConnector accessAPIConnector = new AccessAPIConnector(serviceAccount.getPrivateKey(), accessAPI);
        BigDecimal balance = accessAPIConnector.getAccountBalance(serviceAccount.getFlowAddress());
        Assertions.assertNotNull(balance);
    }
}

File: kotlin-example/src/test/kotlin/org/onflow/examples/kotlin/createAccount/CreateAccountExampleTest.kt
package org.onflow.examples.kotlin.createAccount

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.onflow.examples.kotlin.AccessAPIConnector
import org.onflow.flow.common.test.*
import org.onflow.flow.sdk.FlowAccessApi
import org.onflow.flow.sdk.FlowAddress
import org.onflow.flow.sdk.SignatureAlgorithm
import org.onflow.flow.sdk.bytesToHex
import org.onflow.flow.sdk.crypto.Crypto
import org.onflow.flow.sdk.crypto.PublicKey

@FlowEmulatorProjectTest(flowJsonLocation = "../flow/flow.json")
internal class CreateAccountExampleTest {
    // user key pairs using all supported signing algorithms
    private val userKeyPairs = arrayOf(
        Crypto.generateKeyPair(SignatureAlgorithm.ECDSA_P256),
        Crypto.generateKeyPair(SignatureAlgorithm.ECDSA_SECP256k1)
    )

    // user account addresses
    private val userAccountAddress = arrayOf(
        FlowAddress(""),
        FlowAddress("")
    )

    @FlowServiceAccountCredentials
    lateinit var serviceAccount: TestAccount

    @FlowTestClient
    lateinit var accessAPI: FlowAccessApi

    @FlowTestAccount
    lateinit var otherAccount: TestAccount

    // create an account using the service account
    private fun createUserAccount(userPublicKey: PublicKey): FlowAddress {
        val createAccountExample = CreateAccountExample(serviceAccount.privateKey, accessAPI)
        val account = createAccountExample.createAccount(serviceAccount.flowAddress, userPublicKey)
        return account
    }

    @Test
    fun `Can create an account`() {
        for ((index, address) in userAccountAddress.withIndex()) {
            if (address == FlowAddress("")) {
                // create test accounts
                userAccountAddress[index] = createUserAccount(userKeyPairs[index].public)
            }
        }

        for ((index, address) in userAccountAddress.withIndex()) {
            Assertions.assertNotNull(address)
            // check account key is the expected one
            val accessAPIConnector = AccessAPIConnector(serviceAccount.privateKey, accessAPI)
            val newAccountKey = accessAPIConnector.getAccountKey(address, 0)
            Assertions.assertEquals(userKeyPairs[index].public.hex, newAccountKey.publicKey.bytes.bytesToHex())
        }
    }
    @Test
    fun `Can get an account balance`() {
        val accessAPIConnector = AccessAPIConnector(serviceAccount.privateKey, accessAPI)
        val balance = accessAPIConnector.getAccountBalance(serviceAccount.flowAddress)
        Assertions.assertNotNull(balance)
    }
}
